pkg/cloudlogging/puller_test.go

```diff
@@ -98,7 +98,69 @@ func TestPull(t *testing.T) {
 				t.Errorf("Process(%+v) got result diff (-want, +got): %v", tc.name, diff)
 			}
 
-			if diff := cmp.Diff(tc.wantReq, tc.server.req, protocmp.Transform()); diff != "" {
+			if diff := cmp.Diff(tc.wantReq, tc.server.listReq, protocmp.Transform()); diff != "" {
+				t.Errorf("Process(%+v) got request diff (-want, +got): %v", tc.name, diff)
+			}
+		})
+	}
+}
+
+func TestSteamPull(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name          string
+		server        *fakeServer
+		filter        string
+		wantReq       *loggingpb.TailLogEntriesRequest
+		wantResult    []*loggingpb.LogEntry
+		wantErrSubstr string
+	}{
+		{
+			name:   "success",
```

Comment:
Testing with channel tends to be trickier. You need to 1) check if the channel has received the expected number of logs 2) have timeout because if the code has bug that doesn't return the logs, the test will stuck. A rough example:

```go
var gotLogs []*LogEntry
ch := make(chan *LogEntry)
done := make(chan struct{}, 1)
t.Cleanup(func() {
  close(ch)
  close(done)
})

ctx, cancel := context.WithCancel(context.Background())

// Start a go routine to accept logs
go func() {
  for l := range ch {
    gotLogs = append(gotLogs, l)
    if len(gotLogs) == wantNum {
      done <- struct{}{}
      break;
    }
  }
}

if err := StreamPull(ctx, filter, ch); err != nil {
  // Handle err
}



select {
case <-done:
  // got enough logs
case <-time.After(2*time.Second):
  // timeout
}

// Either way, cancel the context to stop the pull
cancel()

// Compare gotLogs with wantLogs
```

Reply:
I wonder if we can simplify here, just assume there is only log and close the channel right after to signal exit.

Reply:
Could you elaborate? Not sure I follow

Reply:
sth like, close(gotLog) should exit the function.
```
			gotLog := make(chan *loggingpb.LogEntry, 1)
			gotErr := p.StreamPull(ctx, filter, gotLog)
			close(gotLog)
```

Reply:
It's possible that when you `close(gotLog)` the logs haven't entered the channel yet.

Reply:
Even on the same goroutine?

Reply:
Updated.